h2/src/main/org/h2/command/Parser.java

```diff
@@ -7866,9 +7866,34 @@ private Prepared parseSet() {
             readIfEqualOrTo();
             Set command = new Set(session, SetTypes.SCHEMA_SEARCH_PATH);
             ArrayList<String> list = Utils.newSmallArrayList();
-            do {
-                list.add(readAliasIdentifier());
-            } while (readIf(COMMA));
+            if (database.getMode().getEnum() == Mode.ModeEnum.PostgreSQL) {
+                boolean hasPgCatalog = false;
+                do {
+                    // some PG clients will send single-quoted alias
+                    String s = currentTokenType == LITERAL ?
+                            readString() : readAliasIdentifier();
+                    if ("$USER".equalsIgnoreCase(s)) {
+                        continue;
+                    }
+                    if ("PG_CATALOG".equalsIgnoreCase(s)) {
```

Comment:
I think you need to check for exact name here too. The real PostgreSQL does not process its own `pg_catalog` before specified schemas only when `pg_catalog` is specified in the proper case in the search path.

`pgCatalog.equals(s)`

Reply:
I tested on real PG and a schema with upper-case `PG_CATALOG` can be created, which can miss really lower-case `pg_catalog` if use case-insensitive check.

I think `database.sysIdentifier("$USER")` and `database.sysIdentifier("PG_CATALOG")` might work well in this special senario.

Reply:
Of course in can be created in both PostgreSQL and H2. These databases are case-sensitive. (Case sensitivity of H2 is actually configurable, but this feature should not be used in PostgreSQL compatibility mode.) `database.sysIdentifier("PG_CATALOG")` should be fine.

But `$user` should be simply hardcoded in lower case, it's not a schema name and the real PostgreSQL does not treat it in special way in any other case.

Reply:
Fixed. And, `database.sysIdentifier("$USER")` cannot pass test (with -ea VM option) :-P